storage-xray-udp/src/main/java/zipkin2/storage/xray_udp/XRayFormatter.java

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright 2016-2017 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.storage.xray_udp;
+
+import com.squareup.moshi.JsonWriter;
+import java.io.IOException;
+import okio.Buffer;
+
+public final class XRayFormatter {
```

Comment:
Add docs to how you came up with the format.. also make the type not public unless required to be. If required to be, mention why.

Reply:
This is a utility class, basically `brave.SpanCustomizer#tag` is using `<String, String>` we can not pass an exception to it. So I had to prepare JSON to be passed to subsegments. Here is the place which is used https://github.com/openzipkin/zipkin-aws/pull/59/files#diff-5669c8fcacad45292f4bc1d457640d24R243. This is the dirtiest place since it is not using `JsonWriter`. Could you review this part again please? If you are agree with this solution I will come up some documents regarding your notes before you stated.  

Here is the relevant docs on XRay as well: The structure is deeply nested and there is not an easy solution with the current state. 
http://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-errors
I had also decompiled XRay source codes and found that place overly complicated.

That formatter class is not something very nice but it gives some information about exception and also let the user to use another logic.